marlowe-runtime/tx/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs

```diff
@@ -311,11 +311,19 @@ buildApplyInputsConstraintsV1 systemStart eraHistory marloweOutput tipSlot metad
   lift $ unless (invalidBefore' <= tipSlot') $ Left $ ValidityLowerBoundTooHigh tipSlot $ fromCardanoSlotNo invalidBefore'
 
   invalidHereafter' <- lift case invalidHereafter of
-    Nothing -> pure case nextMarloweTimeout contract of
-      Nothing -> maxSafeSlot
-      Just nextTimeout -> case utcTimeToSlotNo' nextTimeout of
-        Right slot -> slot
-        _ -> maxSafeSlot
+    Nothing -> case nextMarloweTimeout contract of                                  -- Find the first timeout.
+      Nothing -> pure maxSafeSlot                                                   -- There is no timeout.
+      Just nextTimeout -> case utcTimeToSlotNo' nextTimeout of                      -- There is a timeout.
+        Right slot -> if slot > invalidBefore'                                      -- Check if next timeout is in the future.
+          then pure slot                                                            -- The next timeout is in the future, but note that this might be an empty interval.
+          else do                                                                   -- The next timeout is in the past.
+            invalidBefore'' <- slotStart invalidBefore'                             -- Find the start time of the validity range.
+            pure case nextMarloweTimeoutAfter invalidBefore'' contract of           -- Find the subsequent timeout, if any.
+              Nothing -> maxSafeSlot                                                -- There is no subsequent future timeout.
+              Just subsequentTimeout -> case utcTimeToSlotNo' subsequentTimeout of  -- There is a subsequent future timeout.
+                Right slot' -> slot'                                                -- The subsequent timeout is before the safe horizon, but note that this might be an empty interval.
+                _ ->  maxSafeSlot                                                   -- The subsequent timeout is beyond the safe horizon.
+        _ -> pure maxSafeSlot                                                       -- The next timeout is in the future, but beyond the safe horizon.
```

Comment:
Is it not equivalent to eliminate the first `nextMarloweTimeout` check like so?

```suggestion
    Nothing -> do                                                                   -- The next timeout is in the past.
      invalidBefore'' <- slotStart invalidBefore'                             -- Find the start time of the validity range.
      pure case nextMarloweTimeoutAfter invalidBefore'' contract of           -- Find the subsequent timeout, if any.
        Nothing -> maxSafeSlot                                                -- There is no subsequent future timeout.
        Just subsequentTimeout -> case utcTimeToSlotNo' subsequentTimeout of  -- There is a subsequent future timeout.
          Right slot' -> slot'                                                -- The subsequent timeout is before the safe horizon, but note that this might be an empty interval.
          _ ->  maxSafeSlot                                                   -- The subsequent timeout is beyond the safe horizon.
```

In which case, instead of adding `nextMarloweTimeoutAfter` as an extra function, you could just add the `invalidBefore :: UTCTime` parameter to `nextMarloweTimeout`.

Reply:
Fixed. Thanks for your insight!